keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,410 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes decoupling the API calls for the preemption from the scheduling cycle, to enhance the scheduling throughput of the scheduling failure scenarios.
+
+## Motivation
+
+The scheduler is basically only one in a cluster,
+and hence scheduling throughput is the crucial metric for the scheduler.
+
+The scheduler schedules Pods one by one within the scheduling cycle, 
+and we basically try to reduce the API calls as much as possible to enhance the scheduling cycle throughput.
+
+The binding cycle is the example for this motivation; 
+1. The scheduling cycle decides where Pod should go to,
+2. At the end of the scheduling cycle, the scheduler reserves the Node within the scheduler's cache so that next scheduling cycle will take the current pod into consideration.
+3. The scheduling cycle ends and the binding cycle starts; the binding cycle is run asynchronously, and the scheduler starts the next scheduling cycle.
+
+This flow allows us to decouple the API call to assign Pod to the Node from the scheduling cycle so that the API call doesn't block the scheduling throughput.
+
+But, we have the similar problem with the preemption; the preemption is run at PostFilter extension point which is the part of the scheduling cycle.
+The preemption has to make some API calls to update Pods' condition and delete Pods after all, which could block the scheduling throughput.
+
+scheduler-perf [actually shows](https://github.com/kubernetes/kubernetes/blob/342da505bdefbd849b808cca3cb76c24a993025f/test/integration/scheduler_perf/config/performance-config.yaml#L641) currently the preemption scenario takes too long time, compared to others.
+
+### Goals
+
+- Improve scheduling throughput when pods require issuing preemptions by making API calls asynchronous
+
+### Non-Goals
+
+- Making the same enhancement for DRA is not a goal of this KEP because it's an under-construction feature yet.
+  - If DRA maintainers want, technically they can along with this KEP. But, at least in this KEP, we don't discuss how.
+
+## Proposal
+
+The preemption plugin makes API calls for the preemption asynchronously after `PostFilter` extension point 
+so that the scheduler can continue to other Pods' scheduling while making API calls for preemption.
+After the preemption goroutine is done, the scheduling for the Pod that triggered the preemption will be retried.
```

Comment:
Does this mean that the DefaultPreemption plugin will be executed outside of PostFilter?
In that case, we may want to add a dedicated extension point.

But, we may be able to consider it as another enhancement or Beta or GA criteria. 
@sanposhiho WDYT?

Reply:
That was the first proposal and I suggested that we keep it internal to the preemption plugin because we don't know of other use cases yet.

Having it as part of the framework can be more maintenance effort.

Reply:
> Having it as part of the framework can be more maintenance effort.

Good point. I'm also thinking too. My primary motivation was keeping consistency in the scheduling cycle.
As far as I know, during the scheduling cycle, the scheduling order is guaranteed (not parallel), and the binding cycle could be parallel.

Since this feature, even though it is part of the scheduling cycle, the order of scheduling will not be guaranteed, right?

But, this is still the alpha stage. So, we may be able to revisit here when we find more clear concerns and use cases.

Reply:
Scheduling order has never been guaranteed in kube-scheduler.

Preemption is already highly asynchronous, given that we have to wait for the pods to actually finish. So this is not a major departure.

Reply:
Yeah, that's true. The order depends on the triggered event,  requeuing and repoped timing and so on. I just wanted to say that the current behavior guarantees one by one scheduling during scheduling cycle.

Reply:
> Preemption is already highly asynchronous, given that we have to wait for the pods to actually finish. So this is not a major departure.

That makes sense. We may be able to consider this feature as a non completely fresh feature since the Preemption already depends on the API call during cycle. This is just improve the Preemption performance.

But, that is a just trivial word play. Thanks for the clarifications.

Reply:
Yeah, so I put the idea of a new extension in the alternative section for now, and we can iterate back when we find it necessary actually. For now, as Aldo suggested, I prefer just doing it with all existing scheduler features.

Reply:
That makes sense. Thanks!

Reply:
Is it possible that two asynchronous preemption go routines end up being executed in the reversed order in the scenario where a higher priority pod2 is assigned to the same node than pod1 is already preemting?

It's clear that the scheduler cache will be always consistent (pod2 will be the nominated one), but in go routines pod1 may theoretically win and override pod2 in the status.nominatedNodeName. Is there a mechanism that would prevent that?

Reply:
As described in "The pod2's scheduling is failed and starts the preemption (pod2 is higher priority than pod1)", it wouldn't be an issue, right? The order of goroutine finishes doesn't matter.
Or, are you talking about another scenario?

Reply:
> The order of goroutine finishes doesn't matter.

Yes, I'm talking about this scenario. It's clear that the scheduler cache will be consistent, but I'm thinking just about the asynchronous go routines. Is pod1 go routine canceled when pod2 takes pod1's place?

If not, isn't it possible that go routine for pod2 finish earlier (sets `status.nominatedNodeName` in apiserver), but pod1 routine finishes later and overrides it incorrectly with pod1? It's probably quite unlikely scenario and the consequences might not be big either, but it still might be possible, if pod1 go routine is not cancelled or at least informed that it should abandon setting `status.nominatedNodeName`.

Reply:
Ah, ignore my comment. The nominated node is written into two different pods, not into a node.